### DISCUSSION TITLE
Firefox Compatibility

### DISCUSSION THREAD

**Discussion by Skilly146** [Tsangares/stoned_ape_tools#9] [Feature Requests]
Just putting this here to have a opened discussion for Firefox compatibility and in case anyone has ideas

### Problems

Right now the main problems preventing a Firefox plugin are one, time, and two, Firefox using a different method to inject code that we don't know how to use. Right now @Tsangares is looking into user script extensions like [Tamper Monkey](https://www.tampermonkey.net/) to try and figure out how they do it

**Comment by Tsangares** [comment]
## Manifests
Firefox is more strict with publishing plugins. Also, Firefox uses manifest version 2.0 whereas chrome uses manifest version 3.0. Making a plugin for both requires making two manifests that submit to the respective web-browser's restrictions.

## Injection Motivation
In order to run the Neptunes Pride plugin, we need acess to the javascript variable `NeptunesPride` which is loaded into the browser from `https://np.ironhelmet.com/scripts/triton/game.js` in the first line:
```
if (!NeptunesPride) {var NeptunesPride = {};}
```

Normally, plugins run in their own scope. In order to access `NeptunesPride` I inject `intel.js` into the footer of the web-page detailed in the file in this repository `worker.js` like this:
```
const add_intel_plugin = () => {
	var s = document.createElement('script');
	s.src = chrome.runtime.getURL('intel.js');
	s.id = "intel";
	s.title = "Stoned Ape Tools v" + chrome.runtime.getManifest().version;
	s.onload = function () {
		this.remove();
	};
	(document.head || document.documentElement).appendChild(s);
	console.log(s.title + " background page.");
}
```

## Injection Dilemma
Firefox's reply was:
> Using a custom CSP to allow remote code execution is not allowed. Please remove all remote hosts from the CSP ‘script-src’ declaration, if any.
Which refers to the code injection.

## Solutions

  1. Retry; maybe I was wrong about the injection begin not allowed by firefox
  2. Transfer data between client and plugin  ([ref](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/Sharing_objects_with_page_scripts))
  3. Embed plugin javascript into webpage ([ref](https://player.support.brightcove.com/coding-topics/pass-data-plugin.html))
      a. `embed_in_page`
